File: api/src/new_phone/schemas/phone_app_config.py
import uuid
from datetime import datetime

from pydantic import BaseModel, Field, model_validator


class PhoneAppConfigResponse(BaseModel):
    model_config = {"from_attributes": True}

    id: uuid.UUID
    tenant_id: uuid.UUID

    # App toggles
    directory_enabled: bool
    voicemail_enabled: bool
    call_history_enabled: bool
    parking_enabled: bool
    queue_dashboard_enabled: bool
    settings_enabled: bool
    page_size: int
    company_name: str | None

    # Phone locale / display
    timezone: str
    language: str
    date_format: str
    time_format: str

    # Security — expose bool flag, never the ciphertext
    has_phone_admin_password: bool = False

    # Branding
    logo_url: str | None
    ringtone: str
    backlight_time: int
    screensaver_type: str

    # Firmware
    firmware_url: str | None

    # Codecs
    codec_priority: str

    # Feature codes
    pickup_code: str
    intercom_code: str
    parking_code: str
    dnd_on_code: str | None
    dnd_off_code: str | None
    fwd_unconditional_code: str | None
    fwd_busy_code: str | None
    fwd_noanswer_code: str | None

    # Network / QoS
    dscp_sip: int
    dscp_rtp: int
    vlan_enabled: bool
    vlan_id: int | None
    vlan_priority: int

    # Action URLs
    action_urls_enabled: bool

    created_at: datetime
    updated_at: datetime

    @model_validator(mode="wrap")
    @classmethod
    def _compute_has_pw(cls, values: object, handler: object) -> "PhoneAppConfigResponse":
        # Extract encrypted_phone_admin_password from ORM object before validation
        has_pw = False
        if hasattr(values, "encrypted_phone_admin_password"):
            has_pw = bool(values.encrypted_phone_admin_password)
        result = handler(values)  # type: ignore[operator]
        result.has_phone_admin_password = has_pw
        return result


class PhoneAppConfigUpdate(BaseModel):
    # App toggles
    directory_enabled: bool | None = None
    voicemail_enabled: bool | None = None
    call_history_enabled: bool | None = None
    parking_enabled: bool | None = None
    queue_dashboard_enabled: bool | None = None
    settings_enabled: bool | None = None
    page_size: int | None = Field(None, ge=5, le=50)
    company_name: str | None = None

    # Phone locale / display
    timezone: str | None = None
    language: str | None = None
    date_format: str | None = None
    time_format: str | None = None

    # Security — accept plaintext, service encrypts before storing
    phone_admin_password: str | None = None

    # Branding
    logo_url: str | None = None
    ringtone: str | None = None
    backlight_time: int | None = Field(None, ge=0, le=1800)
    screensaver_type: str | None = None

    # Firmware
    firmware_url: str | None = None

    # Codecs
    codec_priority: str | None = None

    # Feature codes
    pickup_code: str | None = None
    intercom_code: str | None = None
    parking_code: str | None = None
    dnd_on_code: str | None = None
    dnd_off_code: str | None = None
    fwd_unconditional_code: str | None = None
    fwd_busy_code: str | None = None
    fwd_noanswer_code: str | None = None

    # Network / QoS
    dscp_sip: int | None = Field(None, ge=0, le=63)
    dscp_rtp: int | None = Field(None, ge=0, le=63)
    vlan_enabled: bool | None = None
    vlan_id: int | None = Field(None, ge=1, le=4094)
    vlan_priority: int | None = Field(None, ge=0, le=7)

    # Action URLs
    action_urls_enabled: bool | None = None

File: web/src/api/phone-app-config.ts
import { useQuery, useMutation, useQueryClient } from "@tanstack/react-query"
import { apiClient } from "@/lib/api-client"
import { queryKeys } from "./query-keys"

export interface PhoneAppConfig {
  id: string
  tenant_id: string

  // App toggles
  directory_enabled: boolean
  voicemail_enabled: boolean
  call_history_enabled: boolean
  parking_enabled: boolean
  queue_dashboard_enabled: boolean
  settings_enabled: boolean
  page_size: number
  company_name: string | null

  // Phone locale / display
  timezone: string
  language: string
  date_format: string
  time_format: string

  // Security
  has_phone_admin_password: boolean

  // Branding
  logo_url: string | null
  ringtone: string
  backlight_time: number
  screensaver_type: string

  // Firmware
  firmware_url: string | null

  // Codecs
  codec_priority: string

  // Feature codes
  pickup_code: string
  intercom_code: string
  parking_code: string
  dnd_on_code: string | null
  dnd_off_code: string | null
  fwd_unconditional_code: string | null
  fwd_busy_code: string | null
  fwd_noanswer_code: string | null

  // Network / QoS
  dscp_sip: number
  dscp_rtp: number
  vlan_enabled: boolean
  vlan_id: number | null
  vlan_priority: number

  // Action URLs
  action_urls_enabled: boolean

  created_at: string
  updated_at: string
}

export type PhoneAppConfigUpdate = Partial<
  Omit<PhoneAppConfig, "id" | "tenant_id" | "has_phone_admin_password" | "created_at" | "updated_at">
> & {
  phone_admin_password?: string | null
}

export function usePhoneAppConfig(tenantId: string) {
  return useQuery({
    queryKey: queryKeys.phoneAppConfig.detail(tenantId),
    queryFn: () => apiClient.get<PhoneAppConfig>(`/tenants/${tenantId}/devices/phone-app-config`),
  })
}

export function useUpdatePhoneAppConfig() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: ({ tenantId, data }: { tenantId: string; data: PhoneAppConfigUpdate }) =>
      apiClient.patch<PhoneAppConfig>(`/tenants/${tenantId}/devices/phone-app-config`, data),
    onSuccess: (_data, vars) => {
      qc.invalidateQueries({ queryKey: queryKeys.phoneAppConfig.detail(vars.tenantId) })
    },
  })
}
